### PR TITLE
tls file not exist not return error

### LINKED ISSUE #6608: netutil/tls.go: Incorrect paths to tlsKeyFile and tlsCertFile do not cause a startup error
### Describe the bug

Incorrect paths to `tlsKeyFile` and `tlsCertFile` do not cause a startup error.
Should be fixed here:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/netutil/tls.go#L14

### To Reproduce

run VMAgent with extra args:
```bash
--tls=true --tlsKeyFile=/etc/RandomPath --tlsCertFile=whatever
```
Service successfully starts and listens the 8429 port, but connections to it fail with the non-obvious openssl errors

### Version

up to 1.102.0-rc2

### Logs

_No response_

### Screenshots

_No response_

### Used command-line flags

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by yincongcyincong**
### Describe Your Changes
fix #6608
tls file not exist not return error

### CODE REVIEW

**Comment by jiekun** [comment]
Hi! Thank you for your pull request. Please also note this change to the `CHANGELOG`.

File: lib/netutil/tls.go [hunk 1]
```diff
@@ -8,10 +8,24 @@ import (
...
 
 // GetServerTLSConfig returns TLS config for the server.
 func GetServerTLSConfig(tlsCertFile, tlsKeyFile, tlsMinVersion string, tlsCipherSuites []string) (*tls.Config, error) {
+	if !fs.IsPathExist(tlsCertFile) {
+		return nil, fmt.Errorf("cannot find TLS cert file: %q", tlsCertFile)
+	}
+
+	if !fs.IsPathExist(tlsKeyFile) {
+		return nil, fmt.Errorf("cannot find TLS key file: %q", tlsKeyFile)
+	}
```
Comment (hagen1778) [inline lib/netutil/tls.go:22]: Would ` tls.LoadX509KeyPair` make these checks obsolete?

**Review by hagen1778 — COMMENTED** [review]
From the code it looks like the GetCertificate() will be called only on the first connection to the service. And only then http server will read the certificates. So my concern is: are there cases when tls certs will be mapped to the pod after it has started? @zekker6 @Haleygo ?

I know this happens for promscrape config - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959


Please also see linter warnings in CI.

File: lib/netutil/tls.go (hunk 1 above)
Comment (yincongcyincong) [reply lib/netutil/tls.go:22]: yep, just use tls.LoadX509KeyPair is a better way.

**Comment by zekker6** [comment]
> And only then http server will read the certificates. So my concern is: are there cases when tls certs will be mapped to the pod after it has started? @zekker6 @Haleygo ?

I don't recall any use-cases when this can happen apart from CI jobs like it was mentioned in  #4625.

As far as I can see, logic to ignore an error if certificates are missing was added in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/d5a599badce3ff84c3ef236e1e8c3fb38514f272efd57 in order to make it consistent with the rest of `scrape_config` handling as a part of https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959.

**Comment by Haleygo** [comment]
>So my concern is: are there cases when tls certs will be mapped to the pod after it has started?

No, it shouldn't, files must be mounted in pod when vmagent start.
>I know this happens for promscrape config - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959

In promscrape config, it's mostly for `-dryRun` to check the config validity, see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4625.

**Comment by jiekun** [comment]
Setting aside the comments above, these codes look fine to me. 

Personally, I prefer simpler test assertions, but it's a matter of personal preference and not a blocking issue. WDYT?
```go
	f := func(tlsCertFile, tlsKeyFile string, expectErr bool) {
		t.Helper()
		_, err := GetServerTLSConfig(tlsCertFile, tlsKeyFile, "", []string{})
		if !errors.Is(err, nil) != expectErr {  // same as: if errors.Is(err, nil) == expectErr {
			t.Fatalf("expect err: %v, get error: %v", expectErr, err)
		}
	}
```

**Comment by hagen1778** [comment]
> As far as I can see, logic to ignore an error if certificates are missing was added in https://github.com/VictoriaMetrics/VictoriaMetrics/commit/d5a599badce3ff84c3ef236e1e8c3fb38514f272efd57 in order to make it consistent with the rest of scrape_config handling as a part of https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959.

The commit link from your message doesn't work for me. 
Did you mean this commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/2177675b3403a2aff5adb11bd8bfb597a719daa7 ? If yes, I don't see its alignment with https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959.

**Comment by zekker6** [comment]
@hagen1778 Sorry for the confusion (must have copied wrong link), here is a valid link - https://github.com/VictoriaMetrics/VictoriaMetrics/commit/f03e81c6930f212c6627c60b54720e8026edcfdd
This one is a follow-up for https://github.com/VictoriaMetrics/VictoriaMetrics/commit/e16d3f5639d67ff970975d342aaa276e339e9b0c which did address original #4959 issue.

File: lib/netutil/tls.go [hunk 2]
```diff
@@ -12,6 +12,11 @@ import (
 
 // GetServerTLSConfig returns TLS config for the server.
 func GetServerTLSConfig(tlsCertFile, tlsKeyFile, tlsMinVersion string, tlsCipherSuites []string) (*tls.Config, error) {
+	_, err := tls.LoadX509KeyPair(tlsCertFile, tlsKeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load TLS cert from certFile=%q, keyFile=%q: %w", tlsCertFile, tlsKeyFile, err)
```
[the remark this answers is not in the document]
Comment (Haleygo) [reply lib/netutil/tls.go:17]: Callers already has detailed error wrapper.

[the remark this answers is not in the document]
Comment (yincongcyincong) [reply lib/netutil/tls.go:17]: ok, thanks

**Comment by yincongcyincong** [comment]
> ```go
> f := func(tlsCertFile, tlsKeyFile string, expectErr bool) {
> 		t.Helper()
> 		_, err := GetServerTLSConfig(tlsCertFile, tlsKeyFile, "", []string{})
> 		if !errors.Is(err, nil) != expectErr {  // same as: if errors.Is(err, nil) == expectErr {
> 			t.Fatalf("expect err: %v, get error: %v", expectErr, err)
> 		}
> 	}
> ```

thanks for your review

**Comment by hagen1778** [comment]
According to [commit](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/f03e81c6930f212c6627c60b54720e8026edcfdd), that changed behavior from failing to start with bad certitifcate to check the certificate via callback later, the intention is the following:
> - Make sure that invalid/missing TLS CA file or TLS client certificate files at vmagent startup
  don't prevent from processing the corresponding scrape targets after the file becomes correct,
  without the need to restart vmagent.
  Previously scrape targets with invalid TLS CA file or TLS client certificate files
  were permanently dropped after the first attempt to initialize them, and they didn't
  appear until the next vmagent reload or the next change in other places of the loaded scrape configs.

> - Make sure that TLS CA is properly re-loaded from file after it changes without the need to restart vmagent.
  Previously the old TLS CA was used until vmagent restart.

This decision was made by @valyala to support https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959. Accepting this PR would mean restoring the previous behavior and breaking https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959. So I'd leave this decision to @valyala.

**Comment by Haleygo** [comment]
@yincongcyincong @hagen1778 
I don't think the intention of https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4959 would be broken by this pull request, since it's for tolerating bad TLS files of scrape targets, not vmagent itself. And this pull request only addes TLS files check when starting http server for all the components like vmagent.

**Comment by yincongcyincong** [comment]
@Haleygo hello,bro, does this question have a conclution?

**Review by hagen1778 — APPROVED** [review]
LGTM

**Comment by hagen1778** [comment]
Thanks for the contribution @yincongcyincong !

**Pull request merged**